### PR TITLE
feat: Support JMX exporter

### LINKED ISSUE #635: Support JMX exporter
It would be great if the module could support installing and configuring the [Prometheus JMX exporter](https://github.com/prometheus/jmx_exporter).

This could support the standalone and the preparations for the javaagent deployment and write the configuration file.

I will be contributing this feature.

### PR DESCRIPTION
**Pull request by dploeger**
#### Pull Request (PR) description
This implements support for installing and configuring the prometheus JMX exporter.

#### This Pull Request (PR) fixes the following issues

Fixes #635

### CODE REVIEW

**Comment by dploeger** [comment]
The failing test seems to be something general.

**Comment by dploeger** [comment]
Again, the failed tests seem to not have anything to do with my changes. They seem to time out.

File: data/defaults.yaml [hunk 1]
```diff
@@ -312,3 +312,9 @@ prometheus::grok_exporter::package_ensure: 'latest'
 prometheus::grok_exporter::package_name: 'grok_exporter'
 prometheus::grok_exporter::user: 'grok-exporter'
 prometheus::grok_exporter::version: '1.0.0.RC4'
+prometheus::jmx_exporter::extra_groups: []
```
Comment (bastelfreak) [inline data/defaults.yaml]: please add the data to the exporter class directly, not into defaults.yaml. That's a legacy leftover we need to cleanup.

File: manifests/jmx_exporter.pp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    default     => $manage_service
+  }
+
+  $notify_service = ($restart_on_change and $_manage_service) ? {
+    true    => Service[$service_name],
+    default => undef,
+  }
+
+  file {
+    $config_file_location:
```
Comment (bastelfreak) [inline manifests/jmx_exporter.pp]:
```suggestion
  file { $config_file_location:
```

File: data/defaults.yaml (hunk 1 above)
Comment (dploeger) [reply data/defaults.yaml]: Done

File: manifests/jmx_exporter.pp (hunk 2 above)
Comment (dploeger) [reply manifests/jmx_exporter.pp]: Done

File: manifests/jmx_exporter.pp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+#  Download URL for JMX exporter
+# @param proxy_server
+#  Optional proxy server, with port number if needed. ie: https://example.com:8080
+# @param proxy_type
+#  Optional proxy server type (none|http|https|ftp)
+# @param java_options
+#  Optional options for the JVM of the standalone jmx exporter
+class prometheus::jmx_exporter (
+  String[1] $version,
+  String[1] $config_file_location                            = '/etc/jmx-exporter.yaml',
```
Comment (bastelfreak) [inline manifests/jmx_exporter.pp]:
```suggestion
  Stdlib::Absolutepath $config_file_location                 = '/etc/jmx-exporter.yaml',
```

File: manifests/jmx_exporter.pp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  String[1] $service_name                                    = 'jmx_exporter',
+  String[1] $user                                            = 'jmx_exporter',
+  String[1] $group                                           = 'jmx-exporter',
+  Array $extra_groups                                        = [],
```
Comment (bastelfreak) [inline manifests/jmx_exporter.pp]:
```suggestion
  Array[String[1]] $extra_groups                             = [],
```

File: manifests/jmx_exporter.pp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  String[1] $java_bin_path                                   = '/usr/bin/java',
```
Comment (bastelfreak) [inline manifests/jmx_exporter.pp]:
```suggestion
  Stdlib::Absolutepath $java_bin_path                        = '/usr/bin/java',
```

File: manifests/jmx_exporter.pp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  Enum['javaagent', 'httpserver'] $deployment                = 'httpserver',
+  String[1] $os                                              = downcase($facts['kernel']),
+  String[1] $arch                                            = $prometheus::real_arch,
+  Stdlib::Absolutepath $bin_dir                              = $prometheus::bin_dir,
+  String[1] $config_mode                                     = $prometheus::config_mode,
+  Boolean $restart_on_change                                 = true,
+  Boolean $manage_user                                       = true,
+  Boolean $manage_group                                      = true,
+  Boolean $manage_service                                    = true,
+  Optional[Numeric] $port                                    = undef,
```
Comment (bastelfreak) [inline manifests/jmx_exporter.pp]:
```suggestion
  Optional[Stdlib::Port::Unprivileged] $port                 = undef,
```

**Comment by jacobmw** [comment]
These failed tests look to be more related to the test environment having some strange issues rather than the code itself.  Maybe a forced re-test/push could resolve the issues?

**Comment by bastelfreak** [comment]
I restarted the failed jobs. lets see if they turn green now. You can always do this by your own by closing and reopening a PR.

**Comment by dploeger** [comment]
@bastelfreak Thanks, but apparently the Centos 7 - runs are still failing somehow.

**Comment by bastelfreak** [comment]
CentOS Ci fails because the used docker container doesn't work well on Ubuntu 22.04. In https://github.com/voxpupuli/puppet-prometheus/pull/639 I pin the CI to Ubuntu 20.04. Please rebase after #639 got merged

**Comment by dploeger** [comment]
Done, let's see.

**Comment by dploeger** [comment]
@bastelfreak Still no for centos apparently. :(

**Comment by dploeger** [comment]
But at least only Puppet 6/CentOS 7 now. :)

**Comment by dploeger** [comment]
@bastelfreak Thanks for merging! Do you have an idea when this will be released so we can update our Puppetfile?

**Pull request merged**